org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseDocumentLifeCycleHandler.java

```diff
@@ -390,17 +392,19 @@ public ICompilationUnit handleChanged(DidChangeTextDocumentParams params) {
 
 				Range range = changeEvent.getRange();
 				int length;
-
+				IDocument document = JsonRpcHelpers.toDocument(unit.getBuffer());
+				final int startOffset;
 				if (range != null) {
-					length = changeEvent.getRangeLength().intValue();
+					Position start = range.getStart();
+					startOffset = JsonRpcHelpers.toOffset(document, start.getLine(), start.getCharacter());
+					length = DiagnosticsHelper.getLength(unit, range);
 				} else {
 					// range is optional and if not given, the whole file content is replaced
 					length = unit.getSource().length();
 					range = JDTUtils.toRange(unit, 0, length);
```

Comment:
Also, I may have not seen this the first time, but if you move `startOffset` into the `if` portion (range exists), then we don't need to also define it in the `else`, since it's not used below any longer. So I would just remove the `range = ...` line in the else statement.

Reply:
Good catch, I missed that. Changed it.